llvm/lib/Target/X86/lab3/sredneva_anastasiya/X86SrednevaPass.cpp

```diff
@@ -0,0 +1,79 @@
+#include "X86.h"
+#include "X86InstrInfo.h"
+#include "X86Subtarget.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IntrinsicsX86.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/raw_ostream.h"
+
+#define PASS_NAME "x86-muladd-intrin-pass"
+#define PASS_DESC "x86 Muladd Intrinsic Pass"
+
+using namespace llvm;
+
+namespace {
+class MulAddPass : public MachineFunctionPass {
+public:
+  static char ID;
+  MulAddPass() : MachineFunctionPass(ID) {}
+  bool runOnMachineFunction(MachineFunction &MF) override {
+    SmallVector<MachineInstr *> deletedInstrPtr;
+    bool Changed = false;
+    for (auto &MBB : MF) {
+      for (auto MI = MBB.begin(); MI != MBB.end(); ++MI) {
+        if (MI->getOpcode() == X86::MULPDrr) {
+          Register Reg = MI->getOperand(0).getReg();
+          for (auto NextMI = std::next(MI); NextMI != MBB.end(); ++NextMI) {
+            if (NextMI->getOpcode() == X86::ADDPDrr) {
+              if (NextMI->getOperand(1).getReg() == Reg ||
+                  NextMI->getOperand(2).getReg() == Reg) {
+                bool hasDependency = false;
+                for (auto CheckMI = std::next(NextMI); CheckMI != MBB.end();
+                     ++CheckMI) {
+                  for (unsigned i = 1; i < CheckMI->getNumOperands(); ++i) {
+                    if (CheckMI->getOperand(i).getReg() == Reg &&
+                        NextMI->getOperand(0).getReg() != Reg) {
```

Comment:
Is it a typo or intended check. If the later one then it is independent of inner loop. It is more efficient to check that once in the beginning.